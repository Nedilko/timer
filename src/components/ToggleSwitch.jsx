import { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

const bgLight = `bg-[url('./assets/light.svg')]`
const bgDark = `bg-[url('./assets/dark.svg')]`

ToggleSwitch.propTypes = {
  isOn: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
}

function ToggleSwitch({ onChange, isOn }) {
  return (
    <span
      onClick={onChange}
      className={`dark:bg-gray-6ex00 flex h-5 w-9 cursor-pointer appearance-none rounded-full bg-gray-100 ease-linear ${isOn ? 'bg-left' : 'bg-right'} ${isOn ? bgLight : bgDark} bg-contain bg-[length:20px] bg-no-repeat shadow-sm`}
    >
      <span
        className={`${
          isOn ? 'translate-x-full' : ''
        } m-0.5 h-4 w-4 rounded-full bg-gray-200
      transition duration-75 ease-linear dark:bg-gray-700`}
      ></span>
    </span>
  )
}

export default ToggleSwitch
