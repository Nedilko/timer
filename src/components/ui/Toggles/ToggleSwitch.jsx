import { useState } from "react";
import PropTypes from "prop-types";

const bgLight = `bg-[url('./assets/light.svg')]`;
const bgDark = `bg-[url('./assets/dark.svg')]`;

ToggleSwitch.propTypes = {
  isOn: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool,
};

function ToggleSwitch({ onChange, isOn, isDisabled }) {
  const [toggleOn, setToggleOn] = useState(isOn);
  const handleClick = () => {
    setToggleOn((oldState) => !oldState);
    onChange();
  };
  return (
    <span
      onClick={handleClick}
      className={`dark:bg-gray-6ex00 flex h-5 w-9 cursor-pointer appearance-none rounded-full bg-gray-100 ease-linear ${
        toggleOn ? "bg-left" : "bg-right"
      } ${
        toggleOn ? bgLight : bgDark
      } bg-[length:20px] bg-no-repeat shadow-sm ${
        isDisabled ? "pointer-events-none grayscale" : ""
      }`}
    >
      <span
        className={`${
          toggleOn ? "translate-x-full" : ""
        } m-0.5 h-4 w-4 rounded-full bg-gray-200
transition duration-75 ease-linear dark:bg-gray-700`}
      ></span>
    </span>
  );
}

export default ToggleSwitch;
