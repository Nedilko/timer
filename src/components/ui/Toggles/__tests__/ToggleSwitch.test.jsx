import ToggleSwitch from "../ToggleSwitch";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("ToggleSwitch", () => {
  const onChange = jest.fn();

  beforeEach(() => {
    onChange.mockClear();
  });

  it("should match snapshot", () => {
    render(<ToggleSwitch isOn={true} isEnabled={true} onChange={onChange} />);

    expect(screen.getByRole("checkbox")).toMatchSnapshot();
  });

  it("should have grayscale class when isEnabled is false", () => {
    render(<ToggleSwitch isOn={true} isEnabled={false} onChange={onChange} />);

    expect(screen.getByRole("checkbox")).toHaveClass(
      "pointer-events-none grayscale"
    );
  });

  it("should not have grayscale class when isEnabled", () => {
    render(<ToggleSwitch isOn={true} isEnabled={true} onChange={onChange} />);

    expect(screen.getByRole("checkbox")).not.toHaveClass(
      "pointer-events-none grayscale"
    );
  });

  it("should have bg-left class if isOn is true", () => {
    render(<ToggleSwitch isOn={true} isEnabled={true} onChange={onChange} />);

    expect(screen.getByRole("checkbox")).toHaveClass("bg-left");
    expect(screen.getByRole("checkbox")).toHaveClass(
      "bg-[url('./assets/light.svg')]"
    );
  });

  it("should not have bg-left class if isOn is false", () => {
    render(<ToggleSwitch isOn={false} isEnabled={true} onChange={onChange} />);

    expect(screen.getByRole("checkbox")).toHaveClass("bg-right");
    expect(screen.getByRole("checkbox")).toHaveClass(
      "bg-[url('./assets/dark.svg')]"
    );
  });

  // TODO: ask Alex how to catch span inside

  it("should handle onChange", () => {
    render(<ToggleSwitch isOn={true} isEnabled={false} onChange={onChange} />);

    userEvent.click(screen.getByRole("checkbox"));

    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
