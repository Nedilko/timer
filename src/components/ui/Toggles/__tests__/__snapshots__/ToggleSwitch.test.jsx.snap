// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch should match snapshot 1`] = `
<span
  class="dark:bg-gray-6ex00 flex h-5 w-9 cursor-pointer appearance-none rounded-full bg-gray-100 ease-linear bg-left bg-[url('./assets/light.svg')] bg-[length:20px] bg-no-repeat shadow-sm "
  role="checkbox"
>
  <span
    class="translate-x-full m-0.5 h-4 w-4 rounded-full bg-gray-200
transition duration-75 ease-linear dark:bg-gray-700"
  />
</span>
`;
